modin/pandas/test/test_groupby.py

```diff
@@ -1840,51 +1850,109 @@ def test_handle_as_index(
         agg_func = lambda grp: grp.agg(agg_dict)  # noqa: E731 (lambda assignment)
     else:
         selection = None
 
-    df = pandas.DataFrame(data)
-    external_by_cols = GroupBy.validate_by(df.add_prefix("external_"))
+    # Selecting 'by' columns from both sides of the frame so they located in different partitions
+    internal_by = df.columns[
+        [i if i % 2 == 0 else -i for i in range(internal_by_length)]
+    ].tolist()
+    external_by = external_by_cols[:external_by_length]
 
-    if has_categorical_by:
-        df = df.astype({"a": "category"})
+    pd_by = internal_by + external_by
+    md_by = internal_by + [pd.Series(ser) for ser in external_by]
 
-    internal_by = df.columns[:internal_by_length].tolist()
-    external_by = external_by_cols[:external_by_length]
+    grp_result = pd.DataFrame(df).groupby(md_by, as_index=True)
+    grp_reference = df.groupby(pd_by, as_index=False)
 
-    by = internal_by + external_by
+    agg_result = agg_func(grp_result)
+    agg_reference = agg_func(grp_reference)
+
+    if use_backend_agnostic_method:
+        reset_index, drop, lvls_to_drop, cols_to_drop = GroupBy.handle_as_index(
+            result_cols=agg_result.columns,
+            result_index_names=agg_result.index.names,
+            internal_by_cols=internal_by,
+            by_cols_dtypes=df[internal_by].dtypes.values,
+            by_length=len(md_by),
+            selection=selection,
+            drop=len(internal_by) != 0,
+        )
+
+        if len(lvls_to_drop) > 0:
+            agg_result.index = agg_result.index.droplevel(lvls_to_drop)
+        if len(cols_to_drop) > 0:
+            agg_result = agg_result.drop(columns=cols_to_drop)
+        if reset_index:
+            agg_result = agg_result.reset_index(drop=drop)
+    else:
+        GroupBy.handle_as_index_for_dataframe(
+            result=agg_result,
+            internal_by_cols=internal_by,
+            by_cols_dtypes=df[internal_by].dtypes.values,
+            by_length=len(md_by),
+            selection=selection,
+            drop=len(internal_by) != 0,
+            inplace=True,
+        )
+
+    df_equals(agg_result, agg_reference)
 
-    grp_res = pd.DataFrame(df).groupby(by, as_index=True)
-    grp_ref = df.groupby(by, as_index=False)
 
-    res = agg_func(grp_res)
-    ref = agg_func(grp_ref)
+def test_validate_by():
+    """Test ``modin.core.dataframe.algebra.default2pandas.groupby.GroupBy.validate_by``."""
 
-    drop, lvls_to_drop, cols_to_drop = GroupBy.handle_as_index(
-        result_cols=res.columns,
-        result_index_names=res.index.names,
-        internal_by_cols=internal_by,
-        by_cols_dtypes=df[internal_by].dtypes.values,
-        is_multi_by=len(by) > 1,
-        selection=selection,
-        drop=len(internal_by) != 0,
+    def compare(obj1, obj2):
+        assert not (
+            isinstance(obj1, list) ^ isinstance(obj2, list)
+        ), f"Both objects must be instances of the same type: {type(obj1)} != {type(obj2)}."
```

Comment:
what you're checking and what you're reporting are two different things, pls adjust the message

Reply:
changed the check to the `assert type(obj1) == type(obj2)`